File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        boolean valid = true;
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (isOpening(c)) {
                stack.push(c);
            } else {
                if (!stack.empty() && getOpposite(c) == stack.peek()) {
                    stack.pop();
                } else {
                    valid = false;
                    break;
                }
            }
        }

        if (!stack.empty()) {
            valid = false;
        }

        return valid;
    }

    private boolean isOpening(char c) {
        return c == '(' || c == '{' || c == '[';
    }

    private char getOpposite(char c) {
        switch (c) {
            case ')':
                return '(';
            case '}':
                return '{';
            case ']':
                return '[';
            default:
                return '*';
        }
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def __init__(self):
        self.leftParentheses = ["[", "{", "("]

    def isValid(self, s: str) -> bool:
        stack = []

        for bracket in s:
            if bracket in self.leftParentheses:
                stack.append(bracket)
                continue

            if len(stack) == 0 or stack[-1] != self.get_opening_bracket(bracket):
                return False

            stack.pop()

        return len(stack) == 0

    def get_opening_bracket(self, closing_bracket: str) -> str:
        if closing_bracket == ")":
            return "("
        elif closing_bracket == "}":
            return "{"

        return "["
